paddle/framework/attr_checker.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <functional>
+#include <string>
+#include <vector>
+#include "paddle/framework/op_desc.pb.h"
+
+namespace paddle {
+namespace framework {
+
+class OpDesc;
+
+// check whether a value(attribute) fit a certain limit
+template <typename T>
+class LargerThanChecker {
+ public:
+  LargerThanChecker(T lower_bound) : lower_bound_(lower_bound) {}
+  bool operator()(const T& value) const { return value > lower_bound_; }
+
+ private:
+  T lower_bound_;
+};
+
+// we can provide users more common Checker, like 'LessThanChecker',
+// 'BetweenChecker'...
+
+// check whether a certain attribute fit its limits
+// an attribute can have more than one limits
+template <typename T>
+class TypedAttrChecker {
+  typedef std::function<bool(const T& value)> ValueChecker;
+
+ public:
+  TypedAttrChecker(const std::string& attr_name)
+      : attr_name_(attr_name), has_default_(false) {}
+
+  TypedAttrChecker& LargerThan(const T& lower_bound) {
+    value_checkers_.push_back(LargerThanChecker<T>(lower_bound));
+    return *this;
+  }
+
+  // we can add more common limits, like LessThan(), Between()...
+
+  TypedAttrChecker& SetDefault(const T& default_value) {
+    default_value_ = default_value;
+    has_default_ = true;
+    return *this;
+  }
+
+  // allow users provide their own checker
+  TypedAttrChecker& AddCustomChecker(const ValueChecker& checker) {
+    value_checkers_.push_back(checker);
+    return *this;
+  }
+
+  bool operator()(OpDesc& op_desc) const {
```

Comment:
The `TypedAttrChecker` may need a base class for type erasing.

Also, if each checker stores the result into `op_desc`, then each checker must get that value from `op_desc` again. It could a little bit waste.

Maybe 

```
template <typename T>
T operator()(const OpDesc& op_desc) const {
  T t;
  bool alreadySet = op_desc.Has(attr_name_);
  if (alreadySet) {
    op_desc.Get<T>(attr_name_)
  }
  for (auto& chk : checkers_) {
    chk(&t, alreadySet);
  }
  return t;
}
```